[1.20 regression] narrowing failure with unrelated boolean

**Bug Report**

Mypy 1.20 and master report an incorrect error that seems to be a failure of type narrowing in the following case:

```python
class FancyInteger(int):
    name = 'fancy'

    def __repr__(self):
        return f'{int(self)}!'

    def __str__(self):
        return f'{int(self)}!'


def stringify(value: object) -> str:
    if isinstance(value, (str, int)) or value is None:
        if value.__class__.__str__ == FancyInteger.__str__ and isinstance(value, FancyInteger):
            return f"{value.__class__.__name__}.{value.name}"
    return str(value)
```

[**playground link**](https://mypy-play.net/?mypy=latest&python=3.12&gist=a6f8ba95151fa5e0003273b812d8223e)

This passes in MyPy 1.19.1, and is ostensibly correct code -- this is a minimised version, but the original performs similar checks with `enum.StrEnum`, so not that esoteric.

If the `value.__class__.__str__ == FancyInteger.__str__` comparison is removed, or split to a new `if` statement, the error fails to reproduce.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=a6f8ba95151fa5e0003273b812d8223e

**Expected Behavior**

No error

**Actual Behavior**

```shell
$ uvx mypy==1.20
bug.py:14: error: Item "str" of "str | int" has no attribute "name"  [union-attr]
bug.py:14: error: Item "int" of "str | int" has no attribute "name"  [union-attr]
Found 2 errors in 1 file (checked 1 source file)
$ uvx mypy==1.19.1
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: 1.20 & 1.19.1
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.12 & 3.14

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -6979,12 +6979,10 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
         for expr, expr_type in new_types.items():
             all_mappings.append(self.refine_parent_types(expr, expr_type))
         return and_conditional_maps(
-            new_types,
-            reduce_and_conditional_type_maps(all_mappings, use_meet=True),
-            use_meet=True
+            new_types, reduce_and_conditional_type_maps(all_mappings, use_meet=True), use_meet=True
         )
 
-    def refine_parent_types(self, expr: Expression, expr_type: Type) -> Mapping[Expression, Type]:
+    def refine_parent_types(self, expr: Expression, expr_type: Type) -> TypeMap:
         """Checks if the given expr is a 'lookup operation' into a union and iteratively refines
         the parent types based on the 'expr_type'.
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -6957,10 +6957,6 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
     def propagate_up_typemap_info(self, new_types: TypeMap) -> TypeMap:
         """Attempts refining parent expressions of any MemberExpr or IndexExprs in new_types.
 
-        Specifically, this function accepts two mappings of expression to original types:
-        the original mapping (existing_types), and a new mapping (new_types) intended to
-        update the original.
-
         This function iterates through new_types and attempts to use the information to try
         refining any parent types that happen to be unions.
 
@@ -6979,21 +6975,14 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
 
         We return the newly refined map. This map is guaranteed to be a superset of 'new_types'.
         """
-        output_map = {}
+        all_mappings = []
         for expr, expr_type in new_types.items():
-            # The original inferred type should always be present in the output map, of course
-            output_map[expr] = expr_type
-
-            # Next, try using this information to refine the parent types, if applicable.
-            new_mapping = self.refine_parent_types(expr, expr_type)
-            for parent_expr, proposed_parent_type in new_mapping.items():
-                # We don't try inferring anything if we've already inferred something for
-                # the parent expression.
-                # TODO: Consider picking the narrower type instead of always discarding this?
-                if parent_expr in new_types:
-                    continue
-                output_map[parent_expr] = proposed_parent_type
-        return output_map
+            all_mappings.append(self.refine_parent_types(expr, expr_type))
+        return and_conditional_maps(
+            new_types,
+            reduce_and_conditional_type_maps(all_mappings, use_meet=True),
+            use_meet=True
+        )
 
     def refine_parent_types(self, expr: Expression, expr_type: Type) -> Mapping[Expression, Type]:
         """Checks if the given expr is a 'lookup operation' into a union and iteratively refines
@@ -8744,6 +8733,8 @@ def reduce_and_conditional_type_maps(ms: list[TypeMap], *, use_meet: bool) -> Ty
         return ms[0]
     result = ms[0]
     for m in ms[1:]:
+        if not m:
+            continue  # this is a micro-optimisation
         result = and_conditional_maps(result, m, use_meet=use_meet)
     return result
 

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -6975,12 +6975,10 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
 
         We return the newly refined map. This map is guaranteed to be a superset of 'new_types'.
         """
-        all_mappings = []
+        all_mappings = [new_types]
         for expr, expr_type in new_types.items():
             all_mappings.append(self.refine_parent_types(expr, expr_type))
-        return and_conditional_maps(
-            new_types, reduce_and_conditional_type_maps(all_mappings, use_meet=True), use_meet=True
-        )
+        return reduce_and_conditional_type_maps(all_mappings, use_meet=True)
 
     def refine_parent_types(self, expr: Expression, expr_type: Type) -> TypeMap:
         """Checks if the given expr is a 'lookup operation' into a union and iteratively refines

--- a/test-data/unit/check-isinstance.test
+++ b/test-data/unit/check-isinstance.test
@@ -3070,7 +3070,7 @@ if hasattr(mod, "y"):
 def __getattr__(attr: str) -> str: ...
 [builtins fixtures/module.pyi]
 
-[case testMultipleHasAttr-xfail]
+[case testMultipleHasAttr]
 # flags: --warn-unreachable
 # https://github.com/python/mypy/issues/20596
 from __future__ import annotations

--- a/test-data/unit/check-narrowing.test
+++ b/test-data/unit/check-narrowing.test
@@ -3993,3 +3993,25 @@ def f2(func: Callable[..., T], arg: str) -> T:
         return func(arg)
     return func(arg)
 [builtins fixtures/primitives.pyi]
+
+
+[case testPropagatedParentNarrowingMeet]
+# flags: --strict-equality --warn-unreachable
+from __future__ import annotations
+
+class A:
+    tag: int
+
+class B:
+    tag: int
+    name = "b"
+
+class C:
+    tag: str
+
+def stringify(value: A | B | C) -> str:
+    if isinstance(value.tag, int) and isinstance(value, B):
+        reveal_type(value)  # N: Revealed type is "__main__.B"
+        return value.name
+    return ""
+[builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -3699,9 +3699,9 @@ class B(TypedDict):
     num: int
 
 d: A | B
-match d["tag"]: # E: Match statement has unhandled case for values of type "Literal['b']" \
+match d["tag"]: # E: Match statement has unhandled case for values of type "B" \
                 # N: If match statement is intended to be non-exhaustive, add `case _: pass` \
-                # E: Match statement has unhandled case for values of type "B"
+                # E: Match statement has unhandled case for values of type "Literal['b']"
     case "a":
         reveal_type(d)  # N: Revealed type is "TypedDict('__main__.A', {'tag': Literal['a'], 'name': builtins.str})"
         reveal_type(d["name"])  # N: Revealed type is "builtins.str"

